Fix tabs switching couse scrolling up with YouTube's videos

sfc-gh-tszerszen opened the pull request:
<!--
Before contributing (PLEASE READ!)

⚠️ If your contribution is more than a few lines of code, then prior to starting to code on it please post in the issue saying you want to volunteer, then wait for a positive response. And if there is no issue for it yet, create it first.

This helps make sure:

  1. Two people aren't working on the same thing
  2. This is something Streamlit's maintainers believe should be implemented/fixed
  3. Any API, UI, or deeper architectural changes that need to be implemented have been fully thought through by Streamlit's maintainers
  4. Your time is well spent!

More information in our wiki: https://github.com/streamlit/streamlit/wiki/Contributing
-->

## 📚 Context

_Please describe the project or issue background here_

- What kind of change does this PR introduce?

  - [x] Bugfix
  - [ ] Feature
  - [ ] Refactoring
  - [ ] Other, please describe:

## 🧠 Description of Changes

It's a bug fix for [Issue-#5069](https://github.com/streamlit/streamlit/issues/5069). 

When using an HTML IFrame element with YouTube video with tabs it's initial height is set properly, but after loading of YouTube content it is set to 0, which couses unexpected scrolling up on tab change.

Initially IFrame element has height properly setup, but after YouTube content is loaded React renders Video with width 0, to fix this, when we detect width is 0, we set height to 528, which is default height based on the default streamlit width.

- _Add bullet points summarizing your changes here_

  - [ ] This is a breaking API change
  - [x] This is a visible (user-facing) change

**Revised:**

_Insert screenshot of your updated UI/code here_

**Current:**

_Insert screenshot of existing UI/code here_

## 🧪 Testing Done

- [ ] Screenshots included
- [x] Added/Updated unit tests
- [ ] Added/Updated e2e tests

## 🌐 References

_Does this depend on other work, documents, or tickets?_

- **Issue**: It's fixes YouTube's part of [Issue-#5069](https://github.com/streamlit/streamlit/issues/5069).

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

lukasmasuch commented on frontend/src/components/elements/Video/Video.tsx:
```diff
@@ -72,6 +72,21 @@ export default function Video({ element, width }: VideoProps): ReactElement {
+      window.localStorage.setItem(localStorageIFrameWidthKey, width.toString())
+      initialWidth = width.toString()
+    }
+    if (width === 0 && initialWidth !== undefined && initialWidth !== null) {
```
Hmm, are you sure that it can happen that `width` is `0` here? `width` should actually always be non-zero since it is passed from the parent container (see [here](https://github.com/streamlit/streamlit/blob/891ba59e4c622f4b242707e0dabde9f924e11f22/frontend/src/components/core/Block/ElementNodeRenderer.tsx#L368) and comment [here](https://github.com/streamlit/streamlit/blob/891ba59e4c622f4b242707e0dabde9f924e11f22/frontend/src/components/core/Block/ElementNodeRenderer.tsx#L181)). If it is `0` sometimes, there might be actually something wrong with the container inside the tab 🤔 Or maybe it is set to `0` if it is part of the not-selected tabs?

sfc-gh-tszerszen replied:
@LukasMasuch yes, I'm sure about it, I've created a branch with `console.log` statements which show the behaviour explained in the comment, see [draft-example-of-iframe-width-0](https://github.com/streamlit/streamlit/pull/5137). If you look at Console in the dubug you will see that component width becomes 0

lukasmasuch commented on frontend/src/components/elements/Video/Video.tsx:
```diff
@@ -72,6 +72,21 @@ export default function Video({ element, width }: VideoProps): ReactElement {
+    const localStorageIFrameWidthKey = `iframe-width-${url}`
+    let initialWidth = window.localStorage.getItem(localStorageIFrameWidthKey)
+    if (width !== 0) {
+      window.localStorage.setItem(localStorageIFrameWidthKey, width.toString())
```
I think it would be better here to use `useState` hook to keep a non-zero width in the state. I think `localStorage` is mainly for things we want to persist across sessions.

sfc-gh-tszerszen replied:
@LukasMasuch this issue starts to be really tricky and interesting, I've just changed the `Video` component to use state instead of localStorage, and I get `initial width` set to 0, despite it being set with `setInitialWidth` 🤔

sfc-gh-tszerszen replied:
![image](https://user-images.githubusercontent.com/78742618/184215045-dc0383b6-453b-4ece-a3bf-187f25777c7a.png)

sfc-gh-tszerszen commented:
@LukasMasuch thank you for your help with the PR, can we go with this oneliner? Feel free to ammend this PR and make yourself as an author of the solution 👍

lukasmasuch reviewed (approved):
LGTM 👍 One last thing you might want to check: Is the `videoNode.videoHeight` something that is dynamically calculated based on the available space (width of the parent) or it is just the total size of the raw video format? If it is the latter, it might lead to strange problems if it is used within something like columns

sfc-gh-tszerszen commented:
@LukasMasuch it's fixed value, thank you for noticing this out seems like I've introduced a bug with [PR-#5116](https://github.com/streamlit/streamlit/pull/5116), we should revert that PR or go with this PR's approach:

> videoNode.height = width !== 0 ? width * 0.75 : 528

Current status introduces a bug in which component has huge height but width is much smaller, which makes no sense. I'm sorry for that.

Another solution for this is to use some kind of state saving mechanism, because component get's proper width with first render, and after that it gets width 0. I've used `localStorage` previously, don't know why `useState` didn't work

Thank you for finding this, I've updated this PR with below calculation:

> videoNode.height = width !== 0 ? width * 0.75 : 528

What steps do you think we should take here?

lukasmasuch commented:
I had a quick look. I think setting anything via the `videoNode` will cause some problems since the height and width are not anymore autoselected by the video element. I think we need to pass it via the style attribute below. It needs to be set to `undefined` if the width is not `0`, so the height is automatically selected if the `width` is known, e.g.:
```typescript
  if (type === VideoProto.Type.YOUTUBE_IFRAME) {
    const height = width !== 0 ? width * 0.75 : DEFAULT_HEIGHT

    return (
      <iframe
        title={url}
        src={getYoutubeSrc(url)}
        width={width}
        height={height}
        frameBorder="0"
        allow="autoplay; encrypted-media"
        allowFullScreen
      />
    )
  }

  return (
    <video
      ref={videoRef}
      controls
      src={buildMediaUri(url, getBaseUriParts())}
      className="stVideo"
      style={{ width, height: width === 0 ? DEFAULT_HEIGHT : undefined }}
    />
  )
```

sfc-gh-tszerszen commented:
@LukasMasuch thank you, you've definitely saved the day 🥇 I've introduced solution proposed by you and both YouTube and videos work perfectly fine, without scrolling issue and without strange autoscaling bug introduced by me. If you can, please finally confirm this PR, so we can merge this change and revert previous mistakes done by me 👍

lukasmasuch reviewed (approved):
LGTM 👍

sfc-gh-tszerszen committed:
fix tabs switching couse scrolling up
